res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,884 @@
+
+            // Disconnect pads from current mode
+            this.currentMode.forEachComponent(function(component) {
+                component.disconnect();
+                component.outConnect = false;
+            });
+        }
+
+        // Enable LED of new mode to switch to
+        midi.sendShortMsg(0x94 + this.offset, newMode.ledControl, newMode.colourOn);
+
+        // Connect pads to new mode
+        newMode.forEachComponent(function(component) {
+            component.outConnect = true;
+            component.connect();
+            component.trigger();
+        });
+
+        this.currentMode = newMode;
+    };
+
+    // Start in Hotcue mode
+    this.setPadMode(Prime4.padMode.HOTCUE);
+    midi.sendShortMsg(0x94 + offset, 0x0C, Prime4.rgbCode.whiteDark);
+    midi.sendShortMsg(0x94 + offset, 0x0D, Prime4.rgbCode.whiteDark);
+    midi.sendShortMsg(0x94 + offset, 0x0E, Prime4.rgbCode.whiteDark);
+};
+
+Prime4.PadSection.prototype = Object.create(components.ComponentContainer.prototype);
+
+// Assign mode select buttons in XML file
+Prime4.PadSection.prototype.padModeButtonPressed = function(channel, control, value, _status, _group) {
+    if (value) {
+        this.setPadMode(control);
+    }
+};
+
+// Worry about parameter buttons later
+//Prime4.PadSection.prototype.paramButtonPressed = function(channel, control, value, status, group) {};
+
+// Assign pads mapped in XML file
+Prime4.PadSection.prototype.performancePad = function(channel, control, value, status, group) {
+    const i = (control - 0x0E);
+    this.currentMode.pads[i].input(channel, control, value, status, group);
+};
+
+// HOTCUE MODE
+Prime4.hotcueMode = function(deck, offset) {
+    components.ComponentContainer.call(this);
+    this.ledControl = Prime4.padMode.HOTCUE;
+    this.colourOn = Prime4.rgbCode.blue;
+    this.colourOff = Prime4.rgbCode.whiteDark;
+    this.pads = new components.ComponentContainer();
+    for (let i = 1; i <= 8; i++) {
+        this.pads[i] = new components.HotcueButton({
+            number: i,
+            group: deck.currentDeck,
+            midi: [0x94 + offset, 0x0E + i],
+            colorMapper: Prime4ColorMapper,
+            on: this.colourOn,
+            off: this.colourOff,
+            outConnect: false,
+        });
+    }
+};
+Prime4.hotcueMode.prototype = Object.create(components.ComponentContainer.prototype);
+
+// LOOP MODE
+Prime4.loopMode = function(deck, offset) {
+    components.ComponentContainer.call(this);
+    this.ledControl = Prime4.padMode.LOOP;
+    this.colourOn = Prime4.rgbCode.red;
+    this.colourOff = Prime4.rgbCode.whiteDark;
+    this.PerformancePad = function(n) {
+        this.midi = [0x94 + offset, 0x0E + n];
+        this.number = n;
+        this.outKey = "hotcue_" + this.number + "_enabled";
+        this.colorKey = "hotcue_" + this.number + "_color";
+        components.Button.call(this);
+    };
+    this.PerformancePad.prototype = new components.Button({
```

Comment:
setting things on `this` is a little dangerous because external code can access it and change things unintentionally.
```suggestion
    const PerformancePad = function(n) {
        this.midi = [0x94 + offset, 0x0E + n];
        this.number = n;
        this.outKey = "hotcue_" + this.number + "_enabled";
        this.colorKey = "hotcue_" + this.number + "_color";
        components.Button.call(this);
    };
    PerformancePad.prototype = new components.Button({
```